proposals/route-blinding.md

```diff
@@ -0,0 +1,242 @@
+  intermediate nodes except for their immediate predecessor or successor.
+* Intermediate nodes in the blinded route MUST NOT learn their distance to the recipient `N(r)`.
+* Senders MUST NOT learn the real `node_id`s and `scid`s of the blinded intermediate hops after the
+  introduction point `N(0)`.
+* If `N(r)` creates multiple blinded routes to herself, senders MUST NOT be able to tell that these
+  routes lead to the same recipient (unless this information is leaked by higher layers of the
+  protocol, such as using the same `payment_hash`).
+
+### Encrypted blob
+
+Route blinding introduces a new TLV field to the onion `tlv_payload`: the `encrypted_blob`.
+This blob is used to carry the blinded `scid` to use when forwarding the message and may be
+extended with additional data in the future. It uses ChaCha20-Poly1305 as AEAD scheme.
+
+1. type: 10 (`enctlv`)
+2. data:
+    * [`...*byte`:`enctlv`]
+
+Once decrypted, the content of this encrypted blob is itself a TLV stream that may contain any
+tlv record defined in Bolt 4 (onion TLV namespace).
+
+### Creating a blinded route
+
+`N(r)` performs the following steps to create a blinded route:
+
+```text
+Initialization:
+
+  e(0) <- {0;1}^256
+  E(0) = e(0) * G
+
+Blinding:
+
+  For i = 0 to r-1:
+    ss(i) = H(e(i) * P(i)) = H(k(i) * E(i))         // shared secret known only by N(r) and N(i)
+    B(i) = HMAC256("blinded_node_id", ss(i)) * P(i) // Blinded node_id for N(i), private key known only by N(i)
+    rho(i) = HMAC256("rho", ss(i))                  // Key used to encrypt payload for N(i) by N(r)
+    e(i+1) = H(E(i) || ss(i)) * e(i)                // Ephemeral private key, only known by N(r)
+    E(i+1) = H(E(i) || ss(i)) * E(i)                // NB: N(i) must not learn e(i)
+
+Blinded route:
+  
+  (P(0),fees(0),cltv(0),encrypted_blob(0))
+  (B(1),fees(1),cltv(1),encrypted_blob(1))
+  ...
+  (B(r-1),fees(r-1),cltv(r-1),encrypted_blob(r-1))
+```
+
+Note that this is exactly the same construction as Sphinx, but at each hop we use the shared secret
+to derive a blinded `node_id` for `N(i)` for which the private key will only be known by `N(i)`.
+
+The recipient needs to provide `E(0)` and the blinded route to potential senders.
+The `encrypted_blob(i)` is encrypted with ChaCha20-Poly1305 using the `rho(i)` key, and contains
+the real `short_channel_id` to forward to (and potentially other fields). `E(i)` is included as
+additional authenticated data to detect probing attempts by the sender.
+
+Note that the introduction point uses the real `node_id`, not the blinded one, because the sender
+needs to be able to locate this introduction point and find a route to it. But the sender will send
+`E(0)`, which will allow the introduction point to compute the shared secret and correctly forward.
+
+Note that in the specific case of payments, the recipient can sign the invoice with `e(0)`.
+The sender will recover `E(0)` from the signature so no extra field needs to be added to Bolt 11.
```

Comment:
You may want to prove invoice ownership at a latter time which means you have to keep ephemeral secret key after session expiration, in case of site compromise it may be used to recover blinded_route sessions ?

Reply:
Yeah if you want to be able to prove invoice ownership, we'll need to avoid that trick of signing with the private key. In that case we can simply provide the eph_key in a new field in the invoice, it will probably be worth adding.

Reply:
Added a paragraph mentioning this.